### ISSUE TITLE
Build hooks in dev mode

### ISSUE THREAD

**Issue by SimonBerens** [vuejs/vitepress#2537]
### Describe the bug

Build hooks seem to only run on build. I'm not sure if this is intentional, but it's not documented anywhere and I assumed that they would also run in dev mode. 

Specifically, I want to dynamically add stylesheets based on frontmatter parameters, so I am using `transformHead`, but the `link` tags are not being added.

### Reproduction

This reproduction uses `meta` tags instead of `link` tags. 

https://stackblitz.com/edit/vite-9gbr2k?file=docs%2F.vitepress%2Fconfig.ts

### Expected behavior

`robots` meta tag appears in dev mode

### System Info

```sh
stackblitz
```

### Validations

- [X] Check if you're on the [latest VitePress version](https://github.com/vuejs/vitepress/releases/latest).
- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

**Comment by brc-dd** [comment]
You can use vite's [`transformIndexHtml`](https://vitejs.dev/guide/api-plugin.html#transformindexhtml) to add extra head entries for now 👀 vitepress' [`transformPageData`](https://vitepress.dev/reference/site-config#transformpagedata) might also work. But I haven't tested yet.

**Comment by SimonBerens** [comment]
> You can use vite's [`transformIndexHtml`](https://vitejs.dev/guide/api-plugin.html#transformindexhtml) to add extra head entries for now 👀 vitepress' [`transformPageData`](https://vitepress.dev/reference/site-config#transformpagedata) might also work. But I haven't tested yet.

Thanks for the workaround suggestion! 

An example where I encountered the more general "run a build hook in dev mode" is when I was trying to add an rss feed to my blog, similar to how the vuejs [blog](https://github.com/vuejs/blog/blob/main/.vitepress/genFeed.ts) does it. It seems that I have to build my site for the rss feed to show up, but ideally `buildEnd` is run in dev mode as well so I can prototype my rss feed easily.

**Comment by brc-dd** [comment]
It would not be efficient to generate the feed on every content update 👀  It might be helpful while you're building the rss feed functionality, but after that, if `buildEnd` is run in dev, you'll have to wait for like seconds each time you update even a word in your markdown file.

**Comment by SimonBerens** [comment]
Why is that? I would think it's similar to other `contentLoader` calls. Is `writeFile` that slow?

**Comment by SimonBerens** [comment]
@brc-dd I don't think this should be tagged with `has-workaround` because there is a large set of cases (e.g. rss feed) that cannot be done without having build hooks run in dev mode.
